core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -811,17 +795,30 @@ object IO {
         } yield fiberA.zipWith(fiberAs)(_ :: _)
     }
 
+  final def bracket0[E, A, B](
+    acquire: IO[E, A]
+  )(release: (A, ExitResult[E, B]) => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
+    IORef[E, Option[(A, ExitResult[E, B])]](None).flatMap { m =>
+      (for {
+        a <- acquire.uninterruptibly
+        b <- use(a).run.flatMap(
+              r =>
+                m.write[E](Some((a, r))) *> (r match {
+                  case ExitResult.Completed(b)  => IO.now(b)
+                  case ExitResult.Failed(e)     => fail[E, B](e)
+                  case ExitResult.Terminated(t) => terminate[E, B](t)
+                })
+            )
+      } yield b).ensuring(m.read.flatMap(_.fold(unit[Void]) { case (a, r) => release(a, r) }).uninterruptibly)
+    }
+
   final def bracket[E, A, B](
     acquire: IO[E, A]
-  )(release: ExitResult[E, A] => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
-    IORef[E, Option[ExitResult[E, A]]](None).flatMap { m =>
+  )(release: A => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
+    IORef[E, Option[A]](None).flatMap { m =>
       (for {
-        er <- acquire.run.flatMap(er => m.write[E](Some(er)) *> point(er)).uninterruptibly
-        b <- er match {
-              case ExitResult.Completed(a)  => use(a)
-              case ExitResult.Failed(e)     => fail[E, B](e)
-              case ExitResult.Terminated(t) => terminate[E, B](t)
-            }
+        a <- acquire.flatMap(a => m.write[E](Some(a)).const(a)).uninterruptibly
+        b <- use(a)
       } yield b).ensuring(m.read.flatMap(_.fold(unit[Void])(release(_))).uninterruptibly)
```

Comment:
We should decide if `uninterruptible` semantics are already a part of `ensuring`. I think so, which means we don't need `.uninterruptibly` on the io passed to `ensuring`.